File: src/main/scala/com/cenerino/algorithms/scala/misc/SumDigits.scala
package com.cenerino.algorithms.scala.misc

import scala.annotation.tailrec
import scala.math.abs

/**
 * Sum digits of an integer without converting it to String and splitting the characters.
 */
object SumDigits {

  def calc(num: Int) = {
    @tailrec
    def calc(num: Int, acc: Int): Int = {
      if (num == 0) acc
      else calc(num /10, acc + (num % 10))
    }

    calc(abs(num), 0)
  }
}

File: src/main/java/com/cenerino/algorithms/java/misc/SumDigits.java
package com.cenerino.algorithms.java.misc;

/**
 * Sum digits of an integer without converting it to String and splitting the characters.
 */
public class SumDigits {

    public static int calc(int num) {
        num = Math.abs(num);
        int result = 0;

        while (num > 0) {
            result += num % 10;
            num = num / 10;
        }

        return result;
    }
}
